### DISCUSSION TITLE
Using GhidraScript.createProgram from pyghidra

### DISCUSSION THREAD

**Discussion by bdemick** [NationalSecurityAgency/ghidra#7333] [Q&A] [answered]
In Jython scripts, the `createProgram` function is available, I'm assuming because there is a `GhidraScript` object there. I might just be looking in the wrong place, but where is this exposed via `pyghidra`? I have a kind of chicken-or-egg problem where I would like to construct a program and its memory blocks on-the-fly, but don't want to use `pyghidra.open_program` to invoke (or create) a loader to do it.

I'm using a recent (in the last week or so) build from the master branch, and have `pyghidra` installed in a Python 3.12.0 virtual environment.

**Comment by astrelsky** [comment]
I just checked and `createProgram` is definitely available in a python ghidra script when using PyGhidra. You should be able to do it exactly as you would have using jython.

**Reply by bdemick** [reply]
Sorry if it wasn't clear - I'm attempting to do this via a standalone Python script importing `pyghidra`, not with a script to be run with, say `analyzeHeadless` or the script launcher in the GUI.

**Reply by astrelsky** [reply, marked as answer]
> Sorry if it wasn't clear - I'm attempting to do this via a standalone Python script importing `pyghidra`, not with a script to be run with, say `analyzeHeadless` or the script launcher in the GUI.

Ah I see. This may seem like a lame solution but GhidraScript does have a no argument public constructor.... Nevermind it's abstract. I did a bit of digging because I have outdated docs but try something like

```py
from ghidra.pyghidra import PyGhidraScriptProvider

PyGhidraScriptProvider().getScriptInstance(None, None).createProgram
```

You could have previously done ghidra.python.PythonScript().createProgram but it looks like that class no longer exists.

**Reply by bdemick** [reply]
Thanks! I had tried to go the `GhidraScript` constructor route and ran into that same issue of the class being abstract. Experimenting with your suggestion a bit now. Currently getting: 

```
>>> prog = PyGhidraScriptProvider().getScriptInstance(None, None).createProgram("test", lid)
Traceback (most recent call last):
  File "GhidraScript.java", line 3742, in ghidra.app.script.GhidraScript.createProgram
Exception: Java Exception

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
java.lang.java.lang.NullPointerException: java.lang.NullPointerException: Cannot invoke "ghidra.app.script.GhidraState.getTool()" because "this.state" is null
```

So it looks like there's still a bit of initialization that I need to work out (probably setting up a project and setting calling `set` on the script instance to set the state), but the object from `getScriptInstance()` looks like it has all the expected functions. Much appreciated, this at least gets me unstuck, I think!

**Reply by bdemick** [reply]
This is probably going a little out-of-scope for this discussion, but I'm trying to construct a GhidraState object, which seems like it's a pretty simple class. It looks like JPype's overload matcher is rejecting the call, because I'm passing `None` in most cases:

```
>>> proj = GhidraProject.createProject(tmploc, "test", True)
>>> mystate = GhidraState(None, proj, None, None, None, None)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: No matching overloads found for constructor ghidra.app.script.GhidraState(NoneType,ghidra.base.project.GhidraProject,NoneType,NoneType,NoneType,NoneType), options are:
	public ghidra.app.script.GhidraState(ghidra.app.script.GhidraState)
	public ghidra.app.script.GhidraState(ghidra.framework.plugintool.PluginTool,ghidra.framework.model.Project,ghidra.program.model.listing.Program,ghidra.program.util.ProgramLocation,ghidra.program.util.ProgramSelection,ghidra.program.util.ProgramSelection)
```

However, I see it being used here with `None` args, so I'm not exactly sure what the problem is. I'm admittedly new to JPype, so I'm probably missing some subtlety.
https://github.com/NationalSecurityAgency/ghidra/blob/14527b015e09172bc4a24423fe3bc11391ef6af9/Ghidra/Features/PyGhidra/src/main/py/src/pyghidra/core.py#L172

**Reply by bdemick** [reply]
Derp. Missed the `project.getProject()` call that gets the `Project` from the `GhidraProject`.

**Reply by astrelsky** [reply]
> Derp. Missed the `project.getProject()` call that gets the `Project` from the `GhidraProject`.

Ah ok, glad you got it. The jpype overload messages can be a bit difficult to look at sometimes and it's easy to miss parameters because the fully qualified names are used.

**Reply by bdemick** [reply]
💯 - I had to copy the actual and candidate names out of the error message and line them up directly to spot the issue.

**Reply by astrelsky** [reply]
> 💯 - I had to copy the actual and candidate names out of the error message and line them up directly to spot the issue.

I might submit a pull request tonight changing it to something like `function(SimpleName1, SimpleName2) => function(QualifiedName1, QualifiedName2)` and see how they feel about it.

**Reply by astrelsky** [reply]
I'm lazy and won't do it now but probably will at some point.
